pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,312 @@
+func (manager *Manager) DeleteEgressPolicy(configID policyID) error {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	storedConfig := manager.policyConfigs[configID]
+	if storedConfig == nil {
+		return fmt.Errorf("policy not found")
+	}
+	log.WithFields(logrus.Fields{"policyID": configID}).
+		Debug("Delete local egress policy")
+
+	for endpointId, policies := range manager.policyEndpoints {
+		var newPolicyList []policyID
+		// make a new list excluding policy that is to be deleted
+		for _, policyId := range policies {
+			if policyId == storedConfig.id {
+				// found policy to endpoint mapping, need to delete egress map entry
+				// identified by endpoint and config
+				epData, ok := manager.epDataStore[endpointId]
+				if !ok {
+					return fmt.Errorf("failed to get endpoint data for %v", endpointId)
+				}
+				if err := manager.deleteEgressMap(storedConfig, epData); err != nil {
+					return err
+				}
+			} else {
+				newPolicyList = append(newPolicyList, policyId)
+			}
+		}
+		if len(newPolicyList) > 0 {
+			manager.policyEndpoints[endpointId] = newPolicyList
+		} else {
+			// epDataStore untouched here since endpoint data is unchanged
+			delete(manager.policyEndpoints, endpointId)
+		}
+	}
+	delete(manager.policyConfigs, configID)
+	return nil
+}
+
+// OnUpdateEndpoint is the event handler for endpoint additions and updates.
+func (manager *Manager) OnUpdateEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+	var epData *endpointMetadata
+	var err error
+
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	if len(endpoint.Networking.Addressing) == 0 {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sEndpointName: endpoint.Name,
+			logfields.K8sNamespace:    endpoint.Namespace,
+		}).Error("failed to get valid endpoint IPs, skipping update to egress policy")
+		return
+	}
+
+	if epData, err = manager.getEndpointMetadata(endpoint); err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.K8sEndpointName: endpoint.Name,
+			logfields.K8sNamespace:    endpoint.Namespace,
+		}).Error("failed to get valid endpoint metadata, skipping update to egress policy.")
+		return
+	}
+
+	// Remove old: check if the endpoint was previously selected by any of the policies.
+	if policies, ok := manager.policyEndpoints[epData.id]; ok {
+		for _, policy := range policies {
+			config := manager.policyConfigs[policy]
+			err := manager.deleteEgressMap(config, epData)
+			if err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.K8sEndpointName: endpoint.Name,
+					logfields.K8sNamespace:    endpoint.Namespace,
+				}).Error("Error updating endpoint mapping.")
+				return
+			}
+		}
+		delete(manager.policyEndpoints, epData.id)
+		delete(manager.epDataStore, epData.id)
+	}
+
+	// Upsert new: check if current policies select new endpoint. Also updates endpiont cache
+	manager.epDataStore[epData.id] = epData
```

Comment:
This is a bug in the code. `getEndpointMetadata` had already written the `epData`. I suggest to remove the `manager` receiver from the `getEndpointMetadata` function to avoid updating `manager.epDataStore` by accident.

Reply:
This addition here is necessary because in the previous block of code the ep data might be deleted since it's stale. But you were correct in that the one in getEndpointMetadata is UNnecessary. I think I first wanted to update as early as possible. That one is removed, and tested it does not affect correctness.
Thx!